sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/CachedBatchColumnVector.java

```
@@ -0,0 +1,421 @@
+    super(numRows, DataTypes.NullType, MemoryMode.ON_HEAP);
+    initialize(buffer, type);
+    reserveInternal(numRows);
+    reset();
+  }
+
+  @Override
+  public long valuesNativeAddress() {
+    throw new RuntimeException("Cannot get native address for on heap column");
+  }
+  @Override
+  public long nullsNativeAddress() {
+    throw new RuntimeException("Cannot get native address for on heap column");
+  }
+
+  @Override
+  public void close() {
+  }
+
+  private void setColumnAccessor() {
+    ByteBuffer byteBuffer = ByteBuffer.wrap(buffer);
+    columnAccessor = ColumnAccessor$.MODULE$.apply(type, byteBuffer);
+    calledExtractTo = new boolean[capacity];
+  }
+
+  // call extractTo() before getting actual data
+  private void prepareAccess(int rowId) {
+    if (!calledExtractTo[rowId]) {
+      assert (columnAccessor.hasNext());
+      columnAccessor.extractTo(columnVector, ROWID);
+      calledExtractTo[rowId] = true;
+    }
+  }
+
+  //
+  // APIs dealing with nulls
+  //
+
+  @Override
+  public void putNotNull(int rowId) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void putNull(int rowId) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void putNulls(int rowId, int count) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void putNotNulls(int rowId, int count) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean isNullAt(int rowId) {
+    prepareAccess(rowId);
+    return columnVector.isNullAt(ROWID);
+  }
+
+  //
+  // APIs dealing with Booleans
+  //
+
+  @Override
+  public void putBoolean(int rowId, boolean value) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void putBooleans(int rowId, int count, boolean value) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean getBoolean(int rowId) {
```

then we should throw exception for this case instead of returning wrong result.